contrib/IECoreArnold/src/IECoreArnold/CameraAlgo.cpp

```diff
@@ -147,3 +116,133 @@ AtNode *CameraAlgo::convert( const IECoreScene::Camera *camera, const std::strin
+	// for the normalisation performed in `screenWindow()`.
+	const Imath::Box2f frustum = camera->frustum();
+	const float width = ( frustum.max.x - frustum.min.x ) * 0.5f;
+	return 2.0f * atan( width ) * 180.0f / M_PI;
+}
+
+float apertureSize( const IECoreScene::Camera *camera )
+{
+	if( camera->getFStop() <= 0.0f )
+	{
+		return 0.0f;
+	}
+	// Note the factor of 0.5 because Arnold stores aperture as radius, not diameter.
+	return 0.5f * camera->getFocalLength() * camera->getFocalLengthWorldScale() / camera->getFStop();
+}
+
+template<typename F>
+auto parameterSamples( const std::vector<const IECoreScene::Camera *> &cameraSamples, F &&parameterFunction )
+{
+	using SampleType = typename std::result_of<F( const IECoreScene::Camera *)>::type;
+	std::vector<SampleType> result;
+	result.reserve( cameraSamples.size() );
+	for( const auto &camera : cameraSamples )
+	{
+		result.push_back( parameterFunction( camera ) );
+	}
+	if( std::all_of( result.begin(), result.end(), [&]( const SampleType &x ) { return x == result.front(); } ) )
+	{
+		// If all samples are identical, then deduplicate them down to a single sample.
+		result.resize( 1 );
+	}
+	return result;
+}
+
+template<typename F>
+void setAnimatedFloat( AtNode *node, AtString name, const std::vector<const IECoreScene::Camera *> &cameraSamples, F &&parameterFunction )
+{
+	const auto samples = parameterSamples( cameraSamples, parameterFunction );
+	if( samples.size() > 1 )
+	{
+		AiNodeSetArray( node, name, AiArrayConvert( 1, samples.size(), AI_TYPE_FLOAT, samples.data() ) );
+	}
+	else
+	{
+		AiNodeSetFlt( node, name, samples[0] );
+	}
+}
+
+} // namespace
+
+AtNode *CameraAlgo::convert( const IECoreScene::Camera *camera, const std::string &nodeName, const AtNode *parentNode )
+{
+	AtNode *result = convertCommon( camera, nodeName, parentNode );
+	if( camera->getProjection()=="perspective" )
+	{
+		AiNodeSetFlt( result, g_fovArnoldString, fieldOfView( camera ) );
+		AiNodeSetFlt( result, g_apertureSizeArnoldString, apertureSize( camera ) );
+		AiNodeSetFlt( result, g_focusDistanceArnoldString, camera->getFocusDistance() );
+	}
+
+	const Imath::Box2f sw = screenWindow( camera );
+	AiNodeSetVec2( result, g_screenWindowMinArnoldString, sw.min.x, sw.min.y );
+	AiNodeSetVec2( result, g_screenWindowMaxArnoldString, sw.max.x, sw.max.y );
+
+	return result;
+}
+
+AtNode *CameraAlgo::convert( const std::vector<const IECoreScene::Camera *> &samples, float motionStart, float motionEnd, const std::string &nodeName, const AtNode *parentNode )
+{
+	AtNode *result = convertCommon( samples[0], nodeName, parentNode );
+	if( samples[0]->getProjection()=="perspective" )
+	{
+		setAnimatedFloat( result, g_fovArnoldString, samples, fieldOfView );
+		setAnimatedFloat( result, g_apertureSizeArnoldString, samples, apertureSize );
+		setAnimatedFloat( result, g_focusDistanceArnoldString, samples, []( auto camera ) { return camera->getFocusDistance(); } );
```

Comment:
We've discussed before this style of templating, and I think previously identified that the templating isn't actually doing anything ... in all 3 of these calls, the type being templated on would be resolved to a function pointer that takes a camera and returns a float?  Is this somehow more magical now that you're using a double ampersand in the templates?  Performance doesn't seem crucial here, so it's not a problem if it's not getting inlined, I'm just trying to understand what's happening ... it's a little weird for setAnimatedFloat to be templated if there's only one type it could ever be valid to instantiate that template for ( vs something like parameterSamples that actually works for different types using result_of )

Reply:
Good point. I've pushed a new version which removes the templating from `setAnimatedFloat`, to make that a little clearer. Also spent a bit of time mucking about on godbolt to see what difference it makes. At least in my toy example it seemed to make no difference to what was _possible_, in that the compiler would inline either the templated or non-templated version, and inline the parameter function too if a definition was available. No idea if it makes any difference to the threshold at which it would inline though.